### PR TITLE
Modernize & cleanup

### LINKED ISSUE #40: navigator is not defined
I could  being doing something really stupid (as is the case normally) but I can't get my tests to pass in Node with cuid.

``` bash
error trying to resolve DateInputViewModel navigator is not defined ReferenceError: navigator is not d
efined
```

If I dig a little further, it looks like cuid is expecting the `window.navigator` to exist... but obviously this is not the case when running in a Node process. I am using Testem's launchers to run this particular suite of tests, and perhaps that is where the error lies. Maybe Testem isn't setting an environment variable correctly and therefore some kind of `cuid` check of `if process.env === node` is failing?

Any ideas would be appreciated.

### PR DESCRIPTION
**Pull request by gmaclennan**
Should fix #40 #41 #42 #54 #59 #65 #68 

This is based upon the original source code and api from [`v1.2.5`](https://github.com/ericelliott/cuid/tree/e3c0c4e9b5c1f02ad692b0efcea592e8754c7f36). This is the same code that is published in the latest `v1.3.8`.

I don't know webpack, but this should work fine webpack if you use that, would be great if someone could test it.

I have kept the API from `v1.2.5` and code logic is largely the same, the changes are mainly to restructure the code as commonJS modules, and update code formatting (to match the latest `.eslintrc` in the project (with changes to work with the latest eslint v3.9.1). The code changes are:

- client fingerprints (browser and server) are cached rather than calculated each time, since they do not  change.
- Should work in WebWorkers and ServiceWorkers now - it uses `self` if `window` is undefined.
- Very simple fix for `react-native` which apparently does not have `userAgent` or `mimeTypes` defined. This id does not probably change much between clients, so we should have something with better entropy.

I threw away many of the changes in master, which used ES2016 and required a complicated build process with transforms etc. The project should now be a lot simpler.

Browser builds rely on the [`browser` field in `package.json`](https://github.com/defunctzombie/package-browser-field-spec) which is supported by both browserify and webpack.

Standalone UMD builds are created in `dist` which is excluded from the git repo but included in npm, so that they will be available from https://unpkg.com/cuid/dist/cuid.js https://unpkg.com/cuid/dist/cuid.min.js once this package is published to npm. The build script runs as a prepublish script, so it should require nothing more than `npm publish`.

Tests are running and passing locally and on saucelabs. Hopefully travis-ci should just work.

Finally, sorry about the lack of commit messages, but the repo seemed to have drifted a lot from the last published version and there was no logical step-by-step development from the current `master`.

### CODE REVIEW

**Comment by gmaclennan** [comment]
Looks like saucelabs credentials need updated in `.travis.yml`.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -5,29 +5,23 @@
...
-```
-
-Edit your copy of `bin/env.sh` to enter your Saucelabs username and access key.
+Create a `.zuulrc` file in the root of this project:
 
 ```
-export SAUCE_USERNAME=username
-export SAUCE_ACCESS_KEY=key
+sauce_username: my_awesome_username
+sauce_key: 550e8400-e29b-41d4-a716-446655440000
```
Comment (ericelliott) [inline CONTRIBUTING.md]: If the key comes from `.zuulrc`, which is not included in the repo, and not encrypted - how do we run the sauce labs tests on Travis?

**Comment by ericelliott** [comment]
Notes: 

* https://docs.travis-ci.com/user/pull-requests#Pull-Requests-and-Security-Restrictions
* https://docs.travis-ci.com/user/encryption-keys/
* Sauce Labs has some [dedicated, secure config for sauce connect](https://docs.travis-ci.com/user/sauce-connect/). Any way to make this work with the zuul config?

File: lib/fingerprint.browser.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+var pad = require('./pad.js');
+
+var env = typeof window === 'object' ? window : self;
```
Comment (gmaclennan) [inline lib/fingerprint.browser.js]: `self` is a reference to the global scope (equivalent to `window`) in WebWorkers and ServiceWorkers: https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/self

I looked at a couple of modules designed to work in both a browser context and in WebWorkers and this was the pattern they use.

**Comment by gmaclennan** [comment]
> https://docs.travis-ci.com/user/pull-requests#Pull-Requests-and-Security-Restrictions

Just reading through this. It's likely that the encrypted variables in the current `.travis.yml` are valid saucelabs credentials, it's just that they are not accessible to travis from my PR because it is from a different repo. Hopefully when this is merged to master (or another branch) these tests should just run with those existing credentials.

File: CONTRIBUTING.md (hunk 1 above)
Comment (ericelliott) [inline CONTRIBUTING.md]: I have a Sauce Labs account for this, but as I mentioned in the notes below, it doesn't work for forked PRs.

File: lib/fingerprint.browser.js (hunk 2 above)
Comment (ericelliott) [inline lib/fingerprint.browser.js]: Interesting. TIL.

Can you add tests for Web Workers and Service Workers?

**Comment by ericelliott** [comment]
Any chance we can get a zuul test for the Web Worker/Service Worker fix?

**Comment by gmaclennan** [comment]
I'm not sure this test actually works. I'm not sure whether the TAP output from the worker thread gets back to zuul so that it can confirm the tests are run. This does run the test in a worker, I'm just not sure about how to handle output. I've never run tests in a worker before.

**Comment by gmaclennan** [comment]
A little further info:

When I run zuul locally and run `zuul --local 8080 --ui tape  -- test/test-worker.js` it seems like the test passes, but if I run them both together with a glob it seems to fail: `zuul --local 8080 --ui tape  -- test/**/*.js`

**Comment by ericelliott** [comment]
What do we need to do to get these checks passing so we can close all those open issues?

**Comment by gmaclennan** [comment]
Thanks for following up on this.

The tests are currently failing because of failing saucelab credentials, because this is a PR from a different repo. The only way to fix this it to merge this PR into this repo so that the tests can run with the correct credentials.

I suggest merging this PR into a new branch here, checking the tests pass, then merging into master.

The one thing missing is that I don't think the tests for this working in a webworker are actually working - i.e. if the tests fail in a webworker I don't think this currently cause the test output to show failure. However, this PR has better test coverage than the current version even without this.

**Comment by nealoke** [comment]
@ericelliott any update on this? :)

**Comment by ericelliott** [comment]
@nealoke @gmaclennan I've invited both of you as collaborators. Feel free to do what you need to do in order to get the tests passing on the main repo.

**Comment by nealoke** [comment]
@ericelliott thanks but i'm not really familiar with the CI tools :( When I have time i'll dive into this!

**Comment by MarkHerhold** [comment]
I put a few hours into trying to get this to work with SauceLabs and have not been able to get consistent results. Would it suffice to run the Karma tests on Chrome Headless (v60) and forget the other browsers? Or perhaps try a different testing service?

These are the errors I am seeing:
```
31 08 2017 21:00:10.117:WARN [IE 11.0.0 (Windows 8.1 0.0.0)]: Disconnected (1 times), because no message in 10000 ms.
IE 11.0.0 (Windows 8.1 0.0.0) ERROR
  Disconnected, because no message in 10000 ms.
```

**Comment by ericelliott** [comment]
@MarkHerhold I believe you can adjust the timeout. I'd prefer to make sure it works in IE11, because it currently still holds about 13% market share...

**Comment by ericelliott** [comment]
I think I'd be OK with running the tests in Chrome headless for now just to get this back on track. We could release it as a breaking change, just to be on the safe side. However, I'd want somebody to test it in IE11 and at least spot check it for us... Do you happen to have IE11 available?

**Comment by MarkHerhold** [comment]
@ericelliott I've increased the timeout to 120 seconds ([see this commit on my fork](https://github.com/MarkHerhold/cuid/commit/8c519b8b8eb8ae3610ec7411a73247a7e81d29a9)) and set retries to two and still have flaky results.

Everyone can download an IE11 VM (and other versions) [here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/).

I checked manually and the build does indeed work (tested cuid.js).

image

I will work on getting Chrome Headless working on Travis CI.

**Comment by ericelliott** [comment]
Looking great. Thanks!

**Comment by ericelliott** [comment]
Published as v2.0.1

File: lib/fingerprint.browser.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (zargold) [reply lib/fingerprint.browser.js]: what about node?
```
node
> self
ReferenceError: self is not defined
> window
ReferenceError: window is not defined
>
```

[the remark this answers is not in the document]
Comment (aateeque) [reply lib/fingerprint.browser.js]: What about this point? Was this resolved? I get a 
```
var env = typeof window === 'object' ? window : self;
                                                ^

ReferenceError: self is not defined
```

in node; any suggestions??

**Pull request merged**